util.go

```diff
@@ -50,19 +53,180 @@ func parseTag(path string) string {
 	return strings.TrimSuffix(filepath.Base(path), ".toml")
 }
 
-func parseDependencies(r io.Reader) ([]dependency, error) {
-	var deps []dependency
+func parseDependencies(commit string) ([]dependency, error) {
+	rd, err := fileFromRev(commit, vendorConf)
+	if err == nil {
+		return parseVendorConfDependencies(rd)
+	}
+	rd, err2 := fileFromRev(commit, goMod)
+	if err2 == nil {
+		return parseGoModDependencies(rd)
+	}
+	return nil, errors.Errorf("finding current dep file failed. vendor.conf error: %v, go.mod error: %v", err, err2)
+}
+
+func parseGoModDependencies(r io.Reader) ([]dependency, error) {
+	var err error
+
+	depMap := make(map[string]*dependency)
+	replaceMap := make(map[string]string)
 	s := bufio.NewScanner(r)
 	for s.Scan() {
-		ln := strings.TrimSpace(s.Text())
-		if strings.HasPrefix(ln, "#") || ln == "" {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
 			continue
 		}
-		cidx := strings.Index(ln, "#")
-		if cidx > 0 {
-			ln = ln[:cidx]
+		parts := strings.Fields(ln)
+
+		// scan the file until we find `$DIRECTIVE (`
+		if len(parts) == 2 && parts[1] == "(" {
+			if parts[0] == "require" {
+				depMap, err = processRequiresSection(s)
+				if err != nil {
+					return nil, err
+				}
+			} else if parts[0] == "replace" {
+				replaceMap, err = processReplaceSection(s)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	for depName, version := range replaceMap {
+		if oldDep, ok := depMap[depName]; ok {
+			oldDep.Commit = version
+		} else {
+			logrus.Debugf("dependency %s found in replace section, but doesn't exist in requires section. Skipping", depName)
+			continue
+		}
+	}
+	var deps []dependency
+	for _, dep := range depMap {
+		deps = append(deps, *dep)
+	}
+
+	return deps, nil
+}
+
+func processReplaceSection(s *bufio.Scanner) (map[string]string, error) {
+	replaceMap := make(map[string]string)
+	for s.Scan() {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
+			continue
+		}
+		parts := strings.Fields(ln)
+		numParts := len(parts)
+		if numParts != 4 {
+			if numParts == 1 && parts[0] == ")" {
+				// this is the end of the requires section, break out to process the others
+				break
+			}
+			return nil, fmt.Errorf("invalid config format: %s", ln)
```

Comment:
I would say "unknown" here instead of "invalid", we don't know how this might change in the future, but it is likely just something not yet supported. Throwing a specific type error here that could be caught higher up may be good to suggest updating the tool.

Reply:
fixed, and added a named error, though any handling of that out of the caller